segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java

```diff
@@ -90,29 +87,34 @@ private static Map.Entry<Long, Long> applySearch(StreamSegmentStore store, Strin
 
         return SortUtils.newtonianSearch(idx -> {
             ReadResult result = store.read(indexSegmentName, idx * NameUtils.INDEX_APPEND_EVENT_SIZE, NameUtils.INDEX_APPEND_EVENT_SIZE, TIMEOUT).join();
-            ReadResultEntry firstElement = result.next();
-            // TODO deal with element which is split over multiple entries.
-            switch (firstElement.getType()) {
-                case Cache: // fallthrough
-                case Storage:
-                    return getOffsetFromIndexEntry(firstElement);
-                case Future:
-                    return RETRY_WITH_BACKOFF.retryWhen(e -> true).run(() -> getOffsetFromIndexEntry(firstElement));
-                case Truncated:
-                    throw new SegmentTruncatedException(String.format("Segment %s has been truncated.", segment));
-                case EndOfStreamSegment:
-                    throw new IllegalStateException(String.format("Unexpected size of index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
-            }
-            throw new IllegalStateException(String.format("Unexpected index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
+            result.readRemaining(0, TIMEOUT);
+            return getOffsetFromIndexEntry(indexSegmentName, result);
         }, startIdx, endIdx > 0 ? endIdx - 1 : 0, targetOffset, greater);
     }
 
-    private static long getOffsetFromIndexEntry(ReadResultEntry firstElement) {
-        firstElement.requestContent(TIMEOUT);
-        BufferView content = firstElement.getContent().join();
-        return IndexEntry.fromBytes(content).getOffset();
+    private static long getOffsetFromIndexEntry(String segment, ReadResult readResult) {
+        int bytesRead = 0;
+        ArrayList<BufferView> result = new ArrayList<>(1);
+        while (readResult.hasNext() && bytesRead < NameUtils.INDEX_APPEND_EVENT_SIZE) {
+            ReadResultEntry entry = readResult.next();
```

Comment:
I think  `ReadResultEntry entry = readResult.next();` should be outside while loop. As we are not expecting more than one event in **ReadResult**.